onyo/lib/onyo.py

```diff
@@ -26,12 +27,32 @@ class OnyoProtectedPathError(Exception):
 class Repo:
     """
     """
-    def __init__(self, path: Union[Path, str], init: bool = False) -> None:
+    def __init__(self, path: Union[Path, str], find_root: bool = False,
```

Comment:
What happens in the case both `find_root` and `init` are set to `True`? Since there is no use-case for it, it appears that when `init=False`, we assume `path` is the repo-root, and when `init=True` we assume `path` will become the repo root by virtue of initializing a repository. The only reason `find_root` is here, is to raise an error when it is not True and `path` is not the repo root?

Reply:
I don't think this is resolved yet. The case where `init=True, find_root=True` is functionally equivalent to `init=True, find_root=False`. If both are `True`, it should therefore raise an exception.

Alternatively, we could use `init: Literal['from_path', 'from_root'] = 'from_path'`, because both arguments revolve around how the repository is loaded: (1) from the given path, (2) from the root found from the given path, (3) as a new repository in the given path.

Reply:
Hm. I can understand the viewpoint. But really, forcing the caller to figure out whether `find_root` makes sense, rather than just ignoring it, when it obviously doesn't, is an inconvenience. Don't just think of a caller who is literally passing `True` or `False`. This can be dynamic and relatively complex.

Ultimately, the whole problem is that the instantiation of the `Repo` class and calling `git init` is melted. Datalad started that way, too. In fact: *I* did. Seemed intuitive at first. Turned out to be an ever increasing mistake. But this is a different topic (as in: This particular PR shouldn't care).

Reply:
If we're going to address this topic at a later stage then I'm fine with this use-case, but I'd prefer to be explicit rather than implicit. Right now we're making the assumption that the user realizes `init` and `find_root` are mutually exclusive, and that when using both, only `init` is used and `find_root` is ignored. Such assumptions should generally be avoided.

Reply:
A user could also think that `init` and `find_root` together means: find the root of the git repo above me and init that as an Onyo repo (Onyo allows init-ing a git repo).

I agree this discussion should happen in a different issue/PR. But the intended behavior should be outlined, and then usage made obvious.

Reply:
I do see it the other way around, though, @aqw , @nhjjr :

With this change, the point of `find_root` is to only allow `True` and raise otherwise to "make the user consciously opt into the concept".

And now you are saying, that after being aware of that, I script a use of that, that yes - opts into always `True`. I am again screwed, because always `True` doesn't work. I must only ever call `init=True` with `find_root=False`, but I must never call `init=False` without `find_root=True`.

Instead of just declaring that `find_root` is always implied, unless `init=True` and getting rid of the entire `find_root` parameter, I as the caller not only have to understand the documentation of the function I also need to make my own code more complicated in order to reassure you guys, that I have read the doc.
The only valid and desired behavior is entirely implied by the `init` parameter. To me this feels strange.

Reply:
To be clear: Once I, as the caller, am clear about this. The only thing making sense in calling this dynamically - in my mind- is to say something like `Repo(path, init=initialize_repo, find_root=not intialize_repo)` to make this bearable in my usage.

Reply:
I don't quite get your point, Ben. Let me state first that I'm with Ben and Alex on this topic being for a different PR, but I don't mind having the discussion here for now.

If I understand you correctly, the only reason `find_root` is added in this PR is as an opt-in for an approach that will become the default. If that is the case, we don't need any big discussions about it because ultimately `find_root` will go away and cause no more ambiguity. However, if as a user I provide `init=True, find_root=True` I expect something different to happen from `init=True, find_root=False` but that is not the case. At the very least I should be given a warning that there is no special case when both are True. If your code passes `init=True, find_root=True` when it only expects `init=True` anyway, then the code is confusing. It should be explicit in what it is doing, even if that requires another line of code.

If this is not the case and `find_root` is there to stay along with `init`, I suggest as a better fitting alternative to encapsulate both boolean arguments into one string literal. There are 3 approaches to instantiating `Repo`: (1) initialize a new repo in the given path, (2) use the given path as root, (3) find the root from the given path. 

There is no ambiguity with `load: Literal['from_path', 'from_root', 'init_new'] = 'from_path'` (contents of the string and name of the argument are mere examples). This way there are only 3 options for `load` and no ambiguity with a fourth option that works the same as the third option.

Reply:
> If I understand you correctly, the only reason find_root is added in this PR is as an opt-in for an approach that will become the default.

Default not planned as of now. I was referring to @aqw 's earlier clarification: https://github.com/psyinfra/onyo/pull/334#issuecomment-1525165133

> If as a user I provide init=True, find_root=True I expect something different to happen from init=True, find_root=False but that is not the case

I do understand that, but I don't see raising exceptions as part of "something different to happen" in this context.
With your suggestion there's exactly two valid ways of calling it: `init=True, find_root=False` and `init=False, find_root=True`. Everything else raises. 

> If your code passes init=True, find_root=True when it only expects init=True anyway, then the code is confusing

Well, that is because this function is confusing. It forces me to specify the inevitable. I need to set a parameter with zero degrees of freedom. If it was for `find_root` being ignored in case of `init`, then it would be less of a hassle, because I can still "acknowledge that I have seen the path handling pecularities of it" by always setting it to true.
Which is why I wanted go with that in this PR.

>  It should be explicit in what it is doing, even if that requires another line of code.

The thing is: The calling code doesn't need to explain what it's doing. It's not doing anything other than using the only valid way of instantiating a Repo from a path! Everything else is not about the calling code only about this function's insistence on being told that "Yes, I read the terms and conditions".